lib/Transforms/CombStrengthReduction.cpp

```diff
@@ -0,0 +1,252 @@
+#include "mlir/Pass/Pass.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace circt;
+using namespace comb;
+
+#define GEN_PASS_CLASSES
+#include "circt/Transforms/Passes.h.inc"
+
+static std::pair<size_t, size_t> computeCommonPrefixAndSuffix(const OperandRange &a, const OperandRange &b)
+{
+  size_t commonPrefixLength = 0;
+  size_t commonSuffixLength = 0;
+  size_t sizeA = a.size();
+  size_t sizeB = b.size();
+
+  for (; commonPrefixLength < std::min(sizeA, sizeB); commonPrefixLength++) {
+    if (a[commonPrefixLength] != b[commonPrefixLength]) {
+      break;
+    }
+  }
+
+  for (; commonSuffixLength < std::min(sizeA, sizeB) - commonPrefixLength; commonSuffixLength++) {
+    if (a[sizeA - commonSuffixLength - 1] != b[sizeB - commonSuffixLength - 1]) {
+      break;
+    }
+  }
+
+  return { commonPrefixLength, commonSuffixLength };
+}
+
+static bool isPredicateSigned (const ICmpPredicate & predicate)
+{
+  switch (predicate)  {
+    case ICmpPredicate::ult:
+    case ICmpPredicate::ugt:
+    case ICmpPredicate::ule:
+    case ICmpPredicate::uge:
+    case ICmpPredicate::ne:
+    case ICmpPredicate::eq:
+      return false;
+    case ICmpPredicate::slt:
+    case ICmpPredicate::sgt:
+    case ICmpPredicate::sle:
+    case ICmpPredicate::sge:
+      return true;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+static mlir::Optional<Value> findFirstNonEmptyValue(const OperandRange &range)
+{
+  for (auto op : range) {
+    if (op.getType().getIntOrFloatBitWidth() > 0) {
+      return mlir::Optional<Value>(op);
+    }
+  }
+
+  return mlir::None;
+}
+
+static size_t calculateRangeTotalWidth(const OperandRange &range) {
+  size_t totalWidth = 0;
+  for (auto operand : range) {
+    // getIntOrFloatBitWidth should never raise, since all arguments to ConcatOp are integers.
+    ssize_t width = operand.getType().getIntOrFloatBitWidth();
+    assert (width >= 0);
+    totalWidth += width;
+  }
+  return totalWidth;
+}
+
+LogicalResult matchAndRewriteCompareConcat(ICmpOp &op, ConcatOp &lhs, ConcatOp &rhs, PatternRewriter &rewriter) {
+  auto lhsOperands = lhs.getOperands();
+  auto rhsOperands = rhs.getOperands();
+  size_t numElements = std::min<size_t>(lhsOperands.size(), rhsOperands.size());
+
+  std::pair<size_t, size_t> commonPrefixSuffixLength = computeCommonPrefixAndSuffix(
+      lhsOperands, rhsOperands);
+  size_t commonPrefixLength = commonPrefixSuffixLength.first;
+  size_t commonSuffixLength = commonPrefixSuffixLength.second;
+
+  auto replaceWithConstantI1 = [&](bool constant) -> LogicalResult {
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(op, APInt(1, constant));
+    return success();
+  };
+
+  auto directOrCat = [&](OperandRange range) -> mlir::Value {
```

Comment:
Here you're passing OperandRange by value, other places it is passed by `const &`.  I'm not sure which is right, but please be consistent :)

Reply:
Also, I don't think you need the `mlir::` qualifier on Value.

Reply:
I standardized everything to const OperandRange&, which i think is right. 

We definitely want it as 'const', reference is slightly less copying..